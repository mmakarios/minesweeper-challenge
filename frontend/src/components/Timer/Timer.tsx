import { useState, useEffect, useMemo } from "react";
import styles from "./Timer.module.scss";

interface Props {
  startDate: string;
  endDate?: string;
}

export const Timer = ({ startDate, endDate }: Props) => {
  const [time, setTime] = useState<string>("00:00");

  const initialDate = useMemo(() => {
    return new Date(startDate);
  }, [startDate]);

  const finalDate = useMemo(() => {
    return endDate ? new Date(endDate) : undefined;
  }, [endDate]);

  useEffect(() => {
    const updateTime = (endTime: Date) => {
      const timeDifferenceInSeconds =
        (endTime.getTime() - initialDate.getTime()) / 1000;
      const minutes = Math.floor(timeDifferenceInSeconds / 60).toString();
      const seconds = Math.floor(timeDifferenceInSeconds % 60).toString();
      setTime(
        `${minutes.length <= 1 ? `0${minutes}` : minutes}:${
          seconds.length <= 1 ? `0${seconds}` : seconds
        }`
      );
    };
    if (finalDate) {
      updateTime(finalDate);
      return;
    }
    updateTime(new Date());
    const timer = setInterval(() => updateTime(new Date()), 1000);
    return () => {
      clearInterval(timer);
    };
  }, [initialDate, finalDate]);

  return <div className={styles.timer}>{time}</div>;
};

export default Timer;
