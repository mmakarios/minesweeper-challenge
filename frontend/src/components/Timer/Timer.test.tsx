import React from "react";
import renderer from "react-test-renderer";
import Timer from "./Timer";

it("renders component", () => {
  const tree = renderer.create(<Timer startDate="" />).toJSON();

  expect(tree).toMatchSnapshot();
});
