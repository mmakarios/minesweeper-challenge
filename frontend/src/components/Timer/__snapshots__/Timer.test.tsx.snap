// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders component 1`] = `
<div
  className="timer"
>
  00:00
</div>
`;
